streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -727,6 +727,68 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
+   * elements of the produced streams may be buffered in memory by this operator.
+   */
+  final def flatMapParSwitch[R1 <: R, E1 >: E, B](n: Int, bufferSize: Int = 16)(
+    f: A => ZStream[R1, E1, B]
+  ): ZStream[R1, E1, B] =
+    new ZStream[R1, E1, B] {
+      override def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: Fold[R2, E2, B1, S] =
+        ZManaged.succeedLazy { (s, cont, g) =>
+          for {
+            // Modeled after flatMapPar.
+            out             <- Queue.bounded[Take[E1, B]](bufferSize).toManaged(_.shutdown)
+            permits         <- Semaphore.make(n.toLong).toManaged_
+            innerFailure    <- Promise.make[Cause[E1], Nothing].toManaged_
+            interruptInners <- Promise.make[Nothing, Unit].toManaged_
+            cancelers       <- Queue.bounded[Promise[Nothing, Unit]](n).toManaged(_.shutdown)
+            cancelersGuard  <- Semaphore.make(1).toManaged_
+            driver <- self.foreachManaged { a =>
+                       for {
+                         latch    <- Promise.make[Nothing, Unit]
+                         canceler <- Promise.make[Nothing, Unit]
+                         innerStream = Stream
+                           .bracket(
+                             cancelersGuard.withPermit(
+                               cancelers.size.flatMap { size =>
+                                 if (size < n) UIO.unit
+                                 else cancelers.take.flatMap(_.succeed(()))
+                               } *> cancelers.offer(canceler) *> permits.acquire *> latch.succeed(()).unit
+                             )
+                           )(_ => permits.release)
+                           .flatMap(_ => f(a))
+                           .foreach(b => out.offer(Take.Value(b)).unit)
+                           .foldCauseM(
+                             cause => out.offer(Take.Fail(cause)) *> innerFailure.fail(cause).unit,
+                             _ => UIO.unit
+                           )
+                         _ <- innerStream.raceAll(List(canceler.await, interruptInners.await)).fork
+                         _ <- latch.await
+                       } yield ()
+                     }.foldCauseM(
+                         cause => (interruptInners.succeed(()) *> out.offer(Take.Fail(cause))).unit.toManaged_,
+                         _ =>
+                           innerFailure.await
+                             .raceWith(permits.withPermits(n.toLong)(UIO.unit))(
+                               leftDone =
+                                 (_, permitAcquisition) => interruptInners.succeed(()) *> permitAcquisition.interrupt,
+                               rightDone = (_, failureAwait) => out.offer(Take.End) *> failureAwait.interrupt
+                             )
+                             .toManaged_
+                       )
+                       .fork
+            _ <- ZManaged.finalizer(
```

Comment:
Something I realized after developing `mapMPar` but haven't ported back to `flatMapPar`: using `ZManaged.finalizer` can be unsafe because the computation can be canceled between the previous step and the finalizer step, leading to the finalizer not running.

Let's instead use `.onExit` on the previous step to ensure that a finalizer is installed. And if you can make that change on `flatMapPar` too, that'd be nice :-)

Reply:
Done. Both flatMapPar and flatMapParSwitch.